File: src/backend/app/schemas/academic_record.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from app.schemas.enums import ResultStatus


class CourseAttemptCreate(BaseModel):
    course_code: str
    term_id: str

    attempt_number: int = Field(ge=1, le=2)

    grade: float | None = None

    result_status: ResultStatus

    credits_earned: int = Field(ge=0)

    @model_validator(mode="after")
    def validate_attempt(self):
        if (
            self.result_status
            in {
                ResultStatus.PASSED,
                ResultStatus.FAILED,
            }
            and self.grade is None
        ):
            raise ValueError(
                "Grade is required for Passed or Failed attempts."
            )

        if (
            self.result_status
            in {
                ResultStatus.FAILED,
                ResultStatus.IN_PROGRESS,
            }
            and self.credits_earned != 0
        ):
            raise ValueError(
                "Failed or InProgress attempts "
                "must have 0 earned credits."
            )

        return self


class CourseAttemptResponse(BaseModel):
    attempt_id: str

    course_code: str
    name_vi: str
    name_en: str

    term_id: str
    term_name: str

    attempt_number: int

    grade: float | None

    result_status: ResultStatus

    credits_earned: int


class AcademicRecordResponse(BaseModel):
    record_id: str
    student_id: str
    updated_at: datetime

    earned_credits: int
    passed_courses: list[str]

    attempts: list[CourseAttemptResponse]

File: src/frontend/src/services/AcademicRecord.ts
import { apiRequest } from './api';
import { getCurrentStudentId } from './session';


export type CourseOption = {
  course_code: string;
  course_name: string;
  credits: number;
};


export type AcademicTerm = {
  term_id: string;
  name: string;
};


export type ResultStatus =
  | 'Passed'
  | 'Failed'
  | 'InProgress';


export type CourseAttempt = {
  attempt_id: string;

  course_code: string;
  course_name: string;

  term_id: string;
  term_name: string;

  attempt_number: number;

  grade: number | null;

  result_status: ResultStatus;

  credits_earned: number;
};


export type AcademicRecord = {
  record_id: string;
  student_id: string;
  updated_at: string;

  earned_credits: number;
  passed_courses: string[];

  attempts: CourseAttempt[];
};


export type AttemptCreate = {
  course_code: string;
  term_id: string;

  attempt_number: number;

  grade: number | null;

  result_status: ResultStatus;

  credits_earned: number;
};


export function getCourseOptions() {
  return apiRequest<CourseOption[]>(
    '/api/meta/courses',
  );
}


export function getTerms() {
  return apiRequest<AcademicTerm[]>(
    '/api/meta/terms',
  );
}


export function getAcademicRecord() {
  return apiRequest<AcademicRecord>(
    `/api/students/${getCurrentStudentId()}/academic-record`,
  );
}


export function addAttempt(
  data: AttemptCreate,
) {
  return apiRequest<AcademicRecord>(
    `/api/students/${getCurrentStudentId()}/academic-record/attempts`,
    {
      method: 'POST',
      body: JSON.stringify(data),
    },
  );
}
